### PR TITLE
use direct executor to deflake tests

### PR DESCRIPTION
**Pull request by m-trieu**
`MoreExecutors.directExecutor()/directExecutorService` runs all tasks on the calling thread (w/o offloading to another thread for async work) and calls to `submit` and `execute` will block until the submitted task returns (i.e `Runnable.run()`).

Use this in test implementations of `ChannelCache` and `FanOutStreamingEngineWorkerHarness` to prevent threads waiting on each other.  The old implementation seems to work locally but in the test runner environment has increased in flakiness.

Flakiness is referenced in https://github.com/apache/beam/issues/28957

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCache.java [hunk 1]
```diff
@@ -85,7 +86,9 @@ static ChannelCache forTesting(
         notification -> {
           shutdownChannel(notification.getValue());
           onChannelShutdown.run();
-        });
+        },
+        // Run the removal on the calling thread for better determinism in tests.
```
Comment (Abacn) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCache.java:90]: same here

**Review by Abacn — COMMENTED** [review]
Thanks for the fix. I think we need to understand the test failure was purely testing issue or could also happen in production.

**Comment by m-trieu** [comment]
> Thanks for the fix. I think we need to understand the test failure was purely testing issue or could also happen in production.

This has only shown up in these test suites (haven't run into in load testing).  I wonder if its due to the threads waiting to be scheduled, but the resources are consumed while executing other tests.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/FanOutStreamingEngineWorkerHarness.java [hunk 2]
```diff
@@ -195,7 +199,10 @@ static FanOutStreamingEngineWorkerHarness forTesting(
             getWorkBudgetDistributor,
             dispatcherClient,
             workCommitterFactory,
-            getDataMetricTracker);
+            getDataMetricTracker,
+            // Run the workerMetadataConsumer on the direct calling thread to make testing more
+            // deterministic.
```
[the remark this answers is not in the document]
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/FanOutStreamingEngineWorkerHarness.java:204]: In prod, this is to hand off the task from a thread that (may) perform network IO.  Not needed in testing and can logically be called in line

Added comment.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCache.java (hunk 1 above)
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCache.java:90]: added, this doesn't change any behavior we just want the removal to run synchronously so we don't have to rely on waiting in tests

**Review by Abacn — APPROVED** [review]
Thank you!

**Pull request merged**